distributed/protocol/cupy.py

```diff
@@ -3,6 +3,18 @@
 """
 import cupy
 from .cuda import cuda_serialize, cuda_deserialize
+from distutils.version import LooseVersion
+
+
+class PatchedCudaArrayInterface(object):
+    # TODO: This class wont be necessary
+    #       once Cupy<7.0 is no longer supported
+    def __init__(self, ary):
+        vsn = LooseVersion(cupy.__version__)
+        cai = ary.__cuda_array_interface__
+        if vsn < "7.0.0" and cai.get("strides") is None:
```

Comment:
Maybe makes more sense to check `ary.__cuda_array_interface__["version"] < 2` instead of `vsn < "7.0.0"`? The `__cuda_array_interface__["version"] == 2` is only introduced with CuPy 7.0.0rc1, so versions 7.0.0b4 and lower may not work properly (in case someone is using that).

Reply:
Sure - My only concern here is that we would only want to pop the "strides" attribute if `__cuda_array_interface__["version"] == 2` **and** `vsn < "7.0.0"`.  The problem is that numba supplies a 2nd-version interface, but cupy won't support it until version 7.0.

Reply:
Perhaps we should just check that `vsn < "7.0.0rc1"` (since we only care about the version of the interface if this is the case)?

Reply:
But if we only care about the version of the interface, why don't we check that instead of the version of CuPy? Anyway, either of these checks are _probably_ fine though, unless CuPy backports the `__cuda_array_interface__` which is unlikely.

Reply:
> But if we only care about the version of the interface, why don't we check that instead of the version of CuPy?

I am probably just misunderstanding the suggestion.  However, I don't really see how we can avoid checking the cupy version.  The problem occurs when the interface version is "2", and the cupy version is < 7.0.0rc1 (note that the interface was not defined by a cupy object, but is now being used to create a cupy object).

Reply:
I hope this helps clarify what I mean:

```python
In [1]: import cupy

In [2]: cupy.__version__
Out[2]: '7.0.0rc1'

In [3]: cupy.ones(1000).__cuda_array_interface__["version"]
Out[3]: 2
```

```python
In [1]: import cupy

In [2]: cupy.__version__
Out[2]: '7.0.0b4'

In [3]: cupy.ones(1000).__cuda_array_interface__["version"]
Out[3]: 0
```

Only starting with CuPy 7.0.0rc1 is `__cuda_array_interface__["version"] == 2`, prior to that, it was on version 0. In practice, my suggestion was that the condition would be:

```python
if cay["version"] < 2 and cai.get("strides") is None:
```

But again, the way it's written now is more than fine and probably won't create any issues for anyone, so we can keep it as is.

Reply:
Right - I think I am failing to clarify that the `__cuda_array_interface__` is **not** created by cupy in this case.  The interface is just being used to create a new cupy object.  Therefore, using `if cai["version"] < 2 and cai.get("strides") is None:` will **not** capture the problematic situation, because the interface version **is** 2  (the tests fail with that check in place).

Reply:
I understand now, thanks @rjzamora for the clarification here (and offline). So the issue is that Numba creates the `__cuda_array_interface__` that is later used for Numpy to create the object. Since Numba may be an older version (where `__cuda_array_interface__["version"] < 2"),  the check for CuPy version is needed instead.